pkg/controller/batch/batch_controller.go

```diff
@@ -0,0 +1,392 @@
+		return reconcile.Result{},
+			r.UpdatePhase(v2.BatchFailed, fmt.Errorf("At most one of ['Spec.config', 'spec.ConfigMap'] must be configured"))
+	}
+
+	if spec.ConfigMap == nil {
+		// Create configMap
+		configMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      batch.Name,
+				Namespace: batch.Namespace,
+			},
+		}
+		_, err := controllerutil.CreateOrUpdate(ctx, r.client, configMap, func() error {
+			configMap.Data = map[string]string{BatchFilename: *spec.Config}
+			return controllerutil.SetControllerReference(batch, configMap, r.scheme)
+		})
+
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("Unable to create ConfigMap '%s': %w", configMap.Name, err)
+		}
+
+		updated, err := r.update(func() {
+			batch.Spec.ConfigMap = &batch.Name
+		})
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("Unable to update Batch .Spec: %w", err)
+		}
+		if updated {
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// We update the phase separately to the spec as the status update is ignored when in the update mutate function
+	return reconcile.Result{}, r.UpdatePhase(v2.BatchInitialized, nil)
+}
+
+func (r *batchResource) execute() (reconcile.Result, error) {
+	infinispan, err := r.loadCluster()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	batch := r.instance
+	url, err := connectionUrl(infinispan)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	cliArgs := fmt.Sprintf("--connect '%s' --file '%s/%s'", url, BatchVolumeRoot, BatchFilename)
+
+	labels := batchLabels(batch.Name)
+	infinispan.AddLabelsForPods(labels)
+
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      batch.Name,
+			Namespace: batch.Namespace,
+		},
+		Spec: batchv1.JobSpec{
+			BackoffLimit: pointer.Int32Ptr(0),
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{{
+						Name:    batch.Name,
+						Image:   infinispan.ImageName(),
+						Command: []string{Entrypoint, cliArgs},
+						VolumeMounts: []corev1.VolumeMount{
+							{
+								Name:      BatchVolumeName,
+								MountPath: BatchVolumeRoot,
+							},
+						},
+					}},
+					RestartPolicy: corev1.RestartPolicyNever,
+					Volumes: []corev1.Volume{{
+						Name: BatchVolumeName,
+						VolumeSource: corev1.VolumeSource{
+							ConfigMap: &corev1.ConfigMapVolumeSource{
+								LocalObjectReference: corev1.LocalObjectReference{Name: batch.Name},
+							},
+						},
+					}},
+				},
+			},
+		},
+	}
+
+	_, err = controllerutil.CreateOrUpdate(ctx, r.client, job, func() error {
+		return controllerutil.SetControllerReference(batch, job, r.scheme)
+	})
+
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("Unable to create batch job '%s': %w", batch.Name, err)
+	}
+	return reconcile.Result{}, r.UpdatePhase(v2.BatchRunning, nil)
+}
+
+func (r *batchResource) waitToComplete() (reconcile.Result, error) {
```

Comment:
@ryanemerson could you please a look at this issues #580 described utilizing `dynamic.Interface`
I planned to utilize it for the tests, but probably you can use it in this case

Reply:
Doesn't #580 relate to test code? `waitToComplete()` is only ever called on a reconcile event.

I have updated the code to just return `reconcile.Result{}` instead of `reconcile.Result{RequeueAfter: time.Second}` when the job has not completed, as we already watch Jobs so there's no need to explicitly requeue.

Reply:
It's related to watch for Job conditions without reconciling each not successful stage

Reply:
I'm confused. The batch controller needs to watch for successful and unsuccessful job stages, as a job failure causes the batch CR to fail.

Reply:
Yes, correct. You can use current approach validating on every reconcile loop or use `dynamic.Interface` to watch for changes only. Let's keep as is. Probably we can change it in the future.
But I'm worry about logs length and size placed in to the `Reason` field. Is it too big or not?

Reply:
A failed batch currently only logs exceptions, so the logs are typically very small.

When I first tested this, I actually retrieved the server logs and it worked well (newline chars respected etc) when using `kubectl describe Batch` to view the status.